feat(charts): Adding new chart range selection hook

Abdkhan14 opened the pull request:
This hook generalizes the functionality introduced by: https://github.com/getsentry/sentry/blob/master/static/app/views/explore/hooks/useChartBoxSelect.tsx
<img width="1133" height="403" alt="Screenshot 2025-11-20 at 1 38 06 PM" src="https://github.com/user-attachments/assets/19a09c34-35c7-4a3b-801d-3c226f16164b" />

**Design Notes:**
- Flexible callbacks: `onSelectionStart`, `onSelectionEnd`, and `onClearSelection` let consumers tap into the brushing lifecycle without embedding chart-specific logic in the hook.

- Custom action menu: `actionMenuRenderer` allows callers to fully control the menu’s UI while the hook manages placement, portals, and clear behavior. Passes selected range and utils like `clearSelection` as params

- Controlled redraws: A `deps` array lets callers decide when the selection should reactivate or repaint, keeping multi-chart views in sync.

- Brush mode management: The hook handles enabling brush mode, hiding tooltips during drag, reconnecting chart groups, and restoring visuals as needed.

- Robust outside-click behavior: click handler clears the selection unless the click is inside a marked “safe” region (data-explore-chart-selection-region). This avoids accidental clears while still allowing intuitive dismissal.

**Improvements:**
- Selection UI is butter smooth! 
- Dynamic action menu positioning 

**Added tests**

The cursor comments are valid, I will be addressing them

Commit:
feat(charts): Adding new chart range selection hook

getsantry[bot] committed:
:hammer_and_wrench: apply pre-commit fixes

Commit:
feat(explore-attribute-breakdowns): Addressing cursor comments

Commit:
feat(explore-attr-breakdowns): Fixing eslint errors

edwardgou-sentry commented on static/app/components/charts/useChartXRangeSelection.tsx:
```diff
@@ -0,0 +1,415 @@
+  }, [chartRef, onClearSelection, state?.selection]);
+
+  const onBrushEnd = useCallback<EChartBrushEndHandler>(
+    (evt: any, chartInstance: any) => {
```
Same question about types as `onBrushStart`

Abdkhan14 replied:
Addressed in the comment above.

edwardgou-sentry commented on static/app/components/charts/useChartXRangeSelection.tsx:
```diff
@@ -0,0 +1,415 @@
+      if (tooltipFrameRef.current) cancelAnimationFrame(tooltipFrameRef.current);
+    };
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [state, disabled, enableBrushMode, chartRef.current, chartsGroupName, deps]);
```
Would `chartRef.current` trigger useEffect if it's a ref? Sounds like it might not.
Would it make sense to use `chartRef` instead as a dep to satisfy eslint?

Commit:
feat(charts): Addressing PR comments

edwardgou-sentry reviewed (approved):
I think a couple of the `useEffect`s could use a comment to explain what they are responsible for handling for clarity. Code changes seem reasonable to me though.

Abdkhan14 committed:
feat(explore-attr-breakdowns): Using new useChartXRangeSelection hook (#103814)

- This PR uses the new `useChartXRangeSelection` to enable range
selection only in explore charts
- Removes the old `useBoxChartSelection` hook
<img width="1214" height="663" alt="Screenshot 2025-11-21 at 11 23
04 AM"
src="https://github.com/user-attachments/assets/0c837d4f-0af5-4088-8bbb-75fea46b4832"
/>

---------

Co-authored-by: Abdullah Khan <abdullahkhan@PG9Y57YDXQ.local>
